### PR TITLE
[SDP-1203] Add Circle Metrics

### PR DESCRIPTION
**Pull request by marwen-abid**
### What
* Added Metrics for Circle requests 
* Added optional docker containers for starting Prometheus and Grafana 
* Configured Grafana with the prometheus Data source 
* Removed fixed IDs from the Grafana dashboard and added new panels for monitoring Circle 
* Added Golang Metrics  

#### Circle API Metrics
Screenshot 2024-08-08 at 6 04 26 PM

#### Golang Metrics
Screenshot 2024-08-09 at 11 14 58 AM

### Why
Monitor Requests made to the Circle API and detect any latency issues. 

### Known limitations

[TODO or N/A]

#### PR Structure

* [ ] This PR has a reasonably narrow scope (if not, break it down into smaller PRs).
* [ ] This PR title and description are clear enough for anyone to review it.
* [ ] This PR does not mix refactoring changes with feature changes (split into two PRs otherwise).

#### Thoroughness

* [ ] This PR adds tests for the new functionality or fixes.
* [ ] This PR contains the link to the Jira ticket it addresses.

#### Configs and Secrets

* [ ] No new CONFIG variables are required -OR- the new required ones were added to the helmchart's [`values.yaml`] file.
* [ ] No new CONFIG variables are required -OR- the new required ones were added to the deployments ([`pr-preview`], [`dev`], [`demo`], `prd`).
* [ ] No new SECRETS variables are required -OR- the new required ones were mentioned in the helmchart's [`values.yaml`] file.
* [ ] No new SECRETS variables are required -OR- the new required ones were added to the deployments ([`pr-preview secrets`], [`dev secrets`], [`demo secrets`], `prd secrets`).

#### Release

* [ ] This is not a breaking change.
* [ ] **This is ready for production.**. If your PR is not ready for production, please consider opening additional complementary PRs using this one as the base. Only merge this into `develop` or `main` after it's ready for production!

#### Deployment

* [ ] Does the deployment work after merging?

[`values.yaml`]: ../helmchart/sdp/values.yaml
[`pr-preview`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/stellar-disbursement-platform/backend-helm-values
[`dev`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/backend-helm-values
[`demo`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-backend-helm-values
[`pr-preview secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/externalsecrets-common-previews.yaml#L241-L346
[`dev secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/stellar-disbursement-platform-externalsecrets.yaml
[`demo secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-sdp-externalsecrets.yaml

### CODE REVIEW

File: dev/README.md [hunk 1]
```diff
@@ -286,6 +287,32 @@ Make sure the Docker containers are up and running by executing the `main.sh` sc
...
+
+#### Load the SDP Grafana Dashboard
+
+1. Access the Grafana dashboard by opening a browser and going to [http://localhost:3002](http://localhost:3002).
+2. Log in with the default credentials:
+   - Username: `admin`
+   - Password: `admin`
+3. Click on the `+` icon on the left sidebar and select `Import Dashboard`.
+4. Copy the contents of the [dashboard.json](../resources/grafana/dashboard.json) file and paste it into the `Import via dashboard JSON model` text box.
+
```
Comment (marcelosalloum) [inline dev/README.md:314]: 🤩

File: dev/docker-compose-monitoring.yml
Comment (marcelosalloum) [inline dev/docker-compose-monitoring.yml:1]: 🚀

File: internal/circle/client.go [hunk 2]
```diff
@@ -40,29 +41,31 @@ type ClientInterface interface {
...
+	BasePath       string
+	APIKey         string
+	httpClient     httpclient.HttpClientInterface
+	tenantManager  tenant.ManagerInterface
+	monitorService monitor.MonitorServiceInterface
 }
 
 // ClientFactory is a function that creates a ClientInterface.
-type ClientFactory func(networkType utils.NetworkType, apiKey string, tntManager tenant.ManagerInterface) ClientInterface
+type ClientFactory func(networkType utils.NetworkType, apiKey string, tntManager tenant.ManagerInterface, monitorSvc monitor.MonitorServiceInterface) ClientInterface
```
Comment (marcelosalloum) [inline internal/circle/client.go:52]: We should consider using a struct as the input variable for this method. Maybe for the next change.

File: internal/circle/client.go [hunk 3]
```diff
@@ -282,7 +315,7 @@ func (client *Client) handleError(ctx context.Context, resp *http.Response) erro
 		return fmt.Errorf("parsing API error: %w", err)
 	}
 
-	return fmt.Errorf("Circle API error: %w", apiError) //nolint:golint,unused
+	return fmt.Errorf("circle API error: %w", apiError) //nolint:golint,unused
```
Comment (marcelosalloum) [inline internal/circle/client.go:318]: Do we need the comment `//nolint:golint,unused`?

File: internal/circle/client.go [hunk 4]
```diff
@@ -264,6 +270,33 @@ func parseRetryAfter(retryAfter string) time.Duration {
...
+
+	if monitorErr := client.monitorService.MonitorHistogram(duration.Seconds(), monitor.CircleAPIRequestDurationTag, labels); monitorErr != nil {
+		log.Ctx(ctx).Errorf("monitoring histogram: %v", err)
+	}
+
+	if monitorErr := client.monitorService.MonitorCounters(monitor.CircleAPIRequestsTotalTag, labels); monitorErr != nil {
+		log.Ctx(ctx).Errorf("monitoring counter: %v", err)
+	}
+}
+
```
Comment (marcelosalloum) [inline internal/circle/client.go:299]: Would this make it mandatory to have metrics enabled? I don't think it should be mandatory.

File: internal/circle/service.go [hunk 5]
```diff
@@ -51,6 +54,10 @@ func (o ServiceOptions) Validate() error {
 		return fmt.Errorf("TenantManager is required")
 	}
 
+	if o.MonitorService == nil {
+		return fmt.Errorf("MonitorService is required")
+	}
+
```
Comment (marcelosalloum) [inline internal/circle/service.go:60]: Should monitoring be mandatory?

File: internal/monitor/metric_tags.go
Comment (marcelosalloum) [inline internal/monitor/metric_tags.go:1]: 🤔 maybe we should register failed requests as well, WDYT?

**Review by marcelosalloum — COMMENTED** [review]
Thanks for working on this!

I might be misunderstanding the metrics setup here but am I correct to assume the metrics are being made mandatory?

**Review by marcelosalloum — APPROVED** [review]

File: internal/circle/client.go (hunk 3 above)
Comment (marwen-abid) [reply internal/circle/client.go:318]: I didn't add it, but I'll remove it :)

File: internal/circle/client.go (hunk 4 above)
Comment (marwen-abid) [reply internal/circle/client.go:299]: We discussed this in DMs. Metrics are always collected, (e.g. `sql_exec_with_metrics`) but they can be published or not, based on the flag `METRICS_ENABLED`.

File: internal/circle/service.go (hunk 5 above)
Comment (marwen-abid) [reply internal/circle/service.go:60]: We discussed this in DMs. Metrics are always collected, (e.g. `sql_exec_with_metrics`) but they can be published or not, based on the flag `METRICS_ENABLED`.

File: internal/monitor/metric_tags.go
Comment (marwen-abid) [reply internal/monitor/metric_tags.go:1]: We're already registering failing requests using `recordCircleAPIMetrics`. They're registered under the same tag as successful requests, but I have a label for status and statusCode. status=failed/success and status code is the http status code.

File: internal/circle/client.go (hunk 3 above)
Comment (marwen-abid) [reply internal/circle/client.go]: addressed in a712a9b61f6dafc8344674f3adc55c3525d11617

File: internal/circle/client.go (hunk 2 above)
Comment (marwen-abid) [reply internal/circle/client.go]: addressed in a712a9b61f6dafc8344674f3adc55c3525d11617

**Pull request merged**